tensorflow/python/framework/ops.py

```diff
@@ -513,16 +513,16 @@ def __iter__(self):
     return _TensorIterator(self, shape[0])
 
   def _shape_as_list(self):
-    if self.shape.ndims is not None:
-      return [dim.value for dim in self.shape.dims]
-    else:
+    shape = self._shape_tuple()
+    if shape is None:
       return None
+    return list(shape)
 
   def _shape_tuple(self):
-    shape = self._shape_as_list()
-    if shape is None:
+    try:
+      return tuple(dim.value for dim in self.shape.dims)
```

Comment:
OOC why not check for `None` explicitly instead of try-exceping `TypeError`?

Reply:
To be fair, that's mostly personal preference since I find it more readable and as far as I know ["Easier to ask for forgiveness than permission" (EAFP)](https://docs.python.org/3/glossary.html) is the more common style used in CPython core.

In this case using try/except is around 30% faster in the common case since the previous code checked for `self.shape.ndims-->self.shape.rank-->len(self.shape._dims)` before trying to iterate. Though I am not sure if this matters in practice.

Reply:
I suspect that such optimization will be negligible in a realistic benchmark. A more noticeable optimization (maybe) is to make Python `TensorShape` a thin pybind11 wrapper around `tensorflow::TensorShape`.

This is slightly complicated by the fact that `TensorShape` behavior is configurable, but overall, it should be doable.

Reply:
@lgeiger Can you please check @superbobry's comments and keep us posted. Thanks!

Reply:
I agree with @superbobry that a pybind wrapper around `tensorflow::TensorShape` could potentially speed things up, but this is out of scope of this PR and would need some more investigation to see if this would result in practical performance improvements that would justify the added complexity.

@superbobry Would you prefer me to revert these changes and explicitely check for `self.shape.ndims`?

Reply:
I agree that pybind11 changes are best left out of this PR. However, I don't think this PR alone is worth merging.

Reassigning to @jaingaurav for a second opinion.

Reply:
:+1: Sounds good. This PR was intended as a cleanup and not as a performance improvement anyway, but due other users of `_shape_as_list` I had to reverted the main changes. Feel free to close if necessary.